doc/port-fec/fec-auto-determination.md

```diff
@@ -0,0 +1,187 @@
+| 50         | 2                                 | rs   |
+| 50         | 4                                 | rs   |
+| 50         | 8                                 | rs   |
+| 50         | 16                                | rs   |
+
+Above tables can be defined as JSON file, and be loaded by determine-fec module. Platform can also override the default FEC mapping by providing its own JSON file.
+
+> [!NOTE]
+> If a port has matched FEC entry in both above tables, then prefers FEC entry in first table. (Only exception: For `lane_speed=10, num_lane=1 or 4`, we prefer FEC entry in second table)
+
+
+### Diagram For Different Use Cases
+
+```mermaid
+sequenceDiagram
+    title FEC determination in DPB case
+
+    actor user as User
+    participant dpb_cli as DPB CLI
+    participant determine_fec as determine-fec module
+    participant config_db as CONFIG_DB
+    participant syncd as SYNCD
+
+
+    user ->>+ dpb_cli: do breakout on a port
+    note over dpb_cli: run today's flow per port:
+    dpb_cli ->>+ dpb_cli: generate the config(speed/lanes/etc) for each new port
+
+    note over dpb_cli,determine_fec: run below additional logic per port:
+    dpb_cli ->>+ determine_fec: call API determine_fec(lane_speed, num_lanes) per port
+    determine_fec ->>+ determine_fec: calculate FEC mode based on common rule
+    determine_fec -->>- dpb_cli: return FEC mode
+
+    note over dpb_cli, syncd: run today's flow:
+    dpb_cli ->>+ config_db: Add new ports in PORT table (today's flow), <BR> additionally with proper FEC if determined above
+
+    par
+        config_db -->>- dpb_cli: Done
+    and
+        config_db ->>+ syncd: notify for new port creation
+    end
+
+    dpb_cli -->>- user: Done
+```
+
+```mermaid
+sequenceDiagram
+    title FEC determination in non-DPB case
+
+    actor user as User
+    participant fec_correct_cli as fec-auto-correct CLI <BR> (just a wrapper)
+    participant determine_fec as determine-fec module
+    participant state_db as STATE_DB
+    participant config_db as CONFIG_DB
+    participant syncd as SYNCD
+
+    user ->>+ fec_correct_cli: auto-correct FEC mode for all ports
+    fec_correct_cli ->>+ determine_fec: call API correct_fec_for_all_ports()
+
+    loop every port
+        determine_fec ->>+ state_db: read optics_type from TRANSCEIVER_INFO table
+        state_db -->>- determine_fec: return optics_type
+        determine_fec ->>+ determine_fec: internally call API determine_fec(lane_speed, num_lanes, optics_type) <BR> which calculates FEC mode based on common rule
+    end
+
+    determine_fec ->>+ config_db: update FEC if needed
+
+    par
+        config_db -->>- determine_fec: Done
+    and
+        config_db ->>+ syncd: notify for FEC update
+    end
+
+    determine_fec -->>- fec_correct_cli: Done
+    fec_correct_cli -->>- user: Done
+
+```
+
+> [!NOTE]
+> In the above use cases, automatically determined FEC mode is saved in running config, user needs to do ```config save``` to save it permanently.
```

Comment:
As discussed in the community review, the FEC settings should be added to app_db and not config_db. In addition there should be a global configuration knob to allow xcvrd auto setting FEC. This knob should be disabled by default to honor backward compatibility.

Reply:
@dgsudharsan 
That's correct. We would be adding these two updates to the HLD

Also, we would update this HLD with following as well:
- Remove CLI mention from the HLD and directly mention regular use-case i.e. xcvrd is the consumer of new FEC determination API (mentioned in this HLD)
- Anytime, there is FEC config in config_db, it would supersede any existing or upcoming FEC setting. 
   For this HLD it implies: Apply FEC determination API only if there is no FEC entry/setting in config_db.